Experimental Syndicate Teleporter

13spacemen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Ports the Experimental Syndicate Teleporter, but improved. Thanks to the PRs below.
https://github.com/ParadiseSS13/Paradise/pull/14934
https://github.com/BeeStation/BeeStation-Hornet/pull/3372
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## What Is It?

- The Experimental Syndicate Teleporter is an 8 TC traitor item.
- Teleports you 4-8 units forward in the direction you're facing
- If about to teleport into a wall, safety mechanism teleports tries to teleport you 3 units parallel to the wall. If no safe place, you gib
- Can be forced to teleport with an EMP, however no safety mechanism. Useful for counterplay
- Safety mechanism does not work if user is wearing Bag Of Holding
- If EMPd while on the floor, if mobs are on the same tile, has a chance to teleport them. If no mobs, has a chance to blink out of existence (delete itself)
- Comes with 4 charges, recharges randomly (10% chance every second to gain a charge)
- If you teleport onto another mob, stuns them and does 20 brute damage

## Why It's Good For The Game
Very cool item, risk and reward gameplay, has counterplays, charging is random so it's never a guaranteed teleport you might even have to wait 30+ seconds for a single charge.
I first saw this item on BeeStation and was surprised to find TG didn't have it, since it fits well into SS13
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Videos
Teleporting around with a var-edited teleporter, you can see the safety mechanism and the gib

https://user-images.githubusercontent.com/46101244/175837373-360d9ba8-b105-4378-9a17-3ea23860d87f.mp4

Telefragging someone

https://user-images.githubusercontent.com/46101244/175837384-6c966de1-b981-47a9-8547-8eba85fcfa1a.mp4



## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: Experimental Syndicate Teleporter is now available in the Traitor Uplink for 8 TC. Warning, not as safe as Nanotrasen Teleporters.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

13spacemen committed:
adds the syndicate teleporter to the game
available at uplink for 8 tc
lets user teleport 4-8 tiles forward, recharges randomly, can gib if no safe emergency location or EMPd/wearing bag of holding

private-tristan commented:
I feel like you should be guaranteed at minimum 1 charge per minute before some unlucky lad goes 5+ minutes without a single use out of this thing and bug reports it.

13spacemen committed:
renamed from syndicate teleporter to experimental teleporter
adjusted descriptions to be better written
documented procs and vars
cleaned up code to be more readable
added correct inhands file
balloon alerts for some messages
user gets a message if bag of holding messes up teleport

13spacemen commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,163 @@
+	desc = "A Syndicate reverse-engineered version of the Nanotrasen portable handheld teleporter. It uses bluespace technology to translocate users, but lacks the advanced safety features of its counterpart. Warranty null and void if exposed to an electromagnetic pulse."
+	icon = 'icons/obj/device.dmi'
+	icon_state = "syndi-tele"
+	throwforce = 5
```
The way I see it, it's a crappy syndicate knockoff of the real hand tele, so the construction materials are a lot heavier due to impurities
5 force isn't really a big deal IMO, also if you throw this at someone they can just pick it up and now your 8 TC item is being used by tiders and/or Sec
I can change it if you want though

13spacemen commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,163 @@
+	attempt_teleport(user, FALSE)
+
+/obj/item/teleporter/process()
+	if(prob(10) && charges < max_charges)
```
@MrMelbert 
I think it fits the theme of it being a crappy syndicate knockoff. The item is decently powerful I mean if it's fully charged it lets you zoom across 16-32 tiles, and whoever's chasing you may not have access or equipment to enter the areas you teleported into.
So the recharge being unpredictable is a bit of balance and risk/reward. You breathe a sigh of relief when you're out of charges, on the run, and "the teleporter beeps" shows up in chat

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,180 @@
+	. = ..()
+	if(.)
+		return
+	attempt_teleport(user, FALSE)
```
Named parameter

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,180 @@
+		charges++
+		if(ishuman(loc))
+			var/mob/living/carbon/human/holder = loc
+			to_chat(holder, span_notice("[src] beeps."))
```
```suggestion
			balloon_alert(holder, "teleporter beeps")
```

Mothblocks replied:
This especially needs to be balloon alerted since it is going to be very common and very important.

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,180 @@
+	if(prob(50/severity))
+		if(ishuman(loc))
+			var/mob/living/carbon/human/user = loc
+			to_chat(user, span_danger("[src] buzzes!"))
```
```suggestion
			balloon_alert(user, span_danger("teleporter buzzes"))
```

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,180 @@
+		if(ishuman(loc))
+			var/mob/living/carbon/human/user = loc
+			to_chat(user, span_danger("[src] buzzes!"))
+			attempt_teleport(user, TRUE)
```
Named parameter

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,180 @@
+				if(!teleported_something)
+					teleport_turf.visible_message(span_danger("[src] activates sporadically, teleporting everyone around it!"))
+					teleported_something = TRUE
+				attempt_teleport(mob_on_same_tile, TRUE)
```
Named parameter

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,180 @@
+			var/turf/teleport_turf = get_turf(src)
+			for(var/mob/living/mob_on_same_tile in teleport_turf)
+				if(!teleported_something)
+					teleport_turf.visible_message(span_danger("[src] activates sporadically, teleporting everyone around it!"))
```
Is this necessary? The visual feedback of everyone being teleported is already there, and I'm not sure this is something you'd be able to see so specifically.

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,180 @@
+
+	var/turf/current_location = get_turf(user)
+	var/area/current_area = get_area(current_location)
+	if(!current_location || current_area.area_flags & NOTELEPORT || is_away_level(current_location.z) || is_centcom_level(current_location.z) || is_reserved_level(current_location.z) || !isturf(user.loc))//If turf was not found or they're on z level 2 or >7 which does not currently exist. or if user is not located on a turf
```
Change this to a proc (`malfuctioning`) and split it into multiple calls there.

Also, your comment is both useless and inaccurate (the >2/<7 number)

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,180 @@
+			panic_teleport(user, destination) //We're in a wall, engage emergency parallel teleport.
+		else
+			if(bagholdingcheck)
+				to_chat(user, span_warning("The bluespace interface on your bag of holding interferes with the teleport!"))
```
```suggestion
				balloon_alert(user, "the bag of holding is interfering!")
```

This is important since you shouldn't need to check the chat to know why you aren't teleporting, especially in a pinch.

13spacemen replied:
@Mothblocks This can't be a balloon alert since the user gets gibbed right after you won't be able to see it

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,180 @@
 			itemUser.visible_message(span_suicide("[user] looks even further depressed as they realize they do not have a head...and suddenly dies of shame!"))
 		return (BRUTELOSS)
 
+/obj/item/teleporter
```
Rename this to syndicate_teleporter or something, there's a lot of things this typepath could be.

JohnFulpWillard replied:
Why was this marked as resolved? It's still here

13spacemen replied:
@JohnFulpWillard I mark stuff as resolved whenever I make the change on my laptop so I can keep track, I just haven't pushed it yet

Mothblocks commented on code/__HELPERS/turfs.dm:
```diff
@@ -1,7 +1,7 @@
 ///Returns location. Returns null if no location was found.
-/proc/get_teleport_loc(turf/location,mob/target,distance = 1, density = FALSE, errorx = 0, errory = 0, eoffsetx = 0, eoffsety = 0)
+/proc/get_teleport_loc(turf/location, mob/target, distance = 1, densitycheck = FALSE, closedturfcheck = FALSE, errorx = 0, errory = 0, eoffsetx = 0, eoffsety = 0)
```
snake_case

density_check, not densitycheck
closed_turf_check, not closedturfcheck

13spacemen committed:
snake case in some vars
renamed paths from teleporter to syndicate_teleporter
named params in procs
changed prob() to use DT_PROB()
bolded num of charges on examine
balloon alerts replace some to_chat()
balloon alerts not given to those not holding tele
malfunctioning check is a proc
feedback when emergency teleport happens

Edelor68 commented:
Don't we already have this item in form of the teleporter briefcase? 
I know there are some differences but the role it fulfills is practically the same. 
Is this intended to replace it? 
If not, why do we need this on top of the briefcase?

13spacemen commented:
@Edelor68 It's not a replacement for the briefcase. The briefcase is:
- reliable, no charges system
- can be upgraded with BSRPED
- can teleport anchored structures
- takes time to set up
- can teleport other people and items to and from the launchpad

This teleporter fulfills a different role

Edelor68 commented:
Yes, as I said there are differences. 
The **role** I was talking about though, is that of a short range teleport which both fulfill.
The big difference in utility, as you have pointed out, is that the briefcase can teleport other things than the user, but right now it is mainly just used to teleport the user, he grabs the thing, and teleports out again, which would probably the same use as the syndicate teleporter.
They have different downsides, the briefcase is slow and clunky and the teleporter unreliable, but I still can't see the briefcase ever being used if this is added as the downsides can so easily be played around by just paying attention to your charges and and knowing where you will teleport.
Just imagine what you have to go through to kill the AI for example, to not even talk about steal objectives.
With the briefcase you would have to walk up with it in hand to the AI sat, which would just instantly get security called on you. Then you have set it up and in the walkway around the sat because otherwise it would not reach far enough. If you manage to get in and kill the AI before someone stops you have to teleport back to the same spot and would nearly certainly get caught then, if someone doesn't reach the briefcase  first and just traps you in the satellite.
For the teleporter you can just approach through space, teleport in with 1-2 charges, kill the AI and teleport out somewhere else before people have barely started to move. 
The difference only gets bigger with easier objectives.
Try to go through the same process with armory or any head of staff office.

TL;DR: 
Same role as briefcase, even with differences.
Vastly superior to briefcase and would make it obsolete. Free steal objectives.

13spacemen commented:
@Edelor68 It's unreliable, you can't just approach the AI from space to kill it. The range is 4-8 tiles, you might end up teleporting to a wall and gibbing, or triggering the emergency teleport and being thrown into space again and having to wait for the random chance to gain a charge
As for the armory, you can't carry guns in your backpack so who cares, and again, it's unreliable.
As for head of staff offices, it's much easier to just set up the briefcase launchpad in maint to steal the locker and then smash it open undetected in maint, rather than teleport in with the syndi teleporter which will probably teleport you into the department instead, having to repeatedly teleport to get in (and you can't even open the locker),

Edelor68 commented:
Balance was actually my minor concern. 
As you showed that can be argued about and tweaked if needed. 
My major concern is still that it fulfills the same role as the briefcase and that I don't see why we need both of them.
It is a good item and idea by itself.

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,200 @@
+/obj/item/syndicate_teleporter/proc/malfunctioning(mob/guy_teleporting, turf/current_location, not_holding_tele = FALSE)
+	var/area/current_area = get_area(current_location)
+	. = FALSE
+	if(!current_location)
+		. = TRUE
+	if(current_area.area_flags & NOTELEPORT)
+		. = TRUE
+	if(is_away_level(current_location.z))
+		. = TRUE
+	if(is_centcom_level(current_location.z))
+		. = TRUE
+	if(is_reserved_level(current_location.z))
+		. = TRUE
+	if(!isturf(guy_teleporting.loc))
+		. = TRUE
```
Move all of this to its own proc, and just `return TRUE`.

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,200 @@
+/obj/item/paper/syndicate_teleporter
+	name = "Teleporter Guide"
+	info = {"<b>Instructions on your new prototype teleporter:</b><br>
+	<br>
+	This teleporter will teleport the user 4-8 meters in the direction they are facing.<br>
+	<br>
+	It has 4 charges, and will recharge over time randomly. No, sticking the teleporter into an APC, microwave, or electrified airlock, will not make it charge faster.<br>
+	<br>
+	<b>Warning:</b> Teleporting into walls will activate a failsafe teleport parallel up to 3 meters, but the user will be ripped apart if it fails to find a safe location.<br>
+	<br>
+	Do not expose the teleporter to electromagnetic pulses. Unwanted malfunctions may occur.
+"}
```
Indent all by one, than leave a new line afterwards

Mothblocks commented on code/game/objects/items/teleportation.dm:
```diff
@@ -309,5 +309,200 @@
+
+/obj/item/syndicate_teleporter/emp_act(severity)
+	var/teleported_something = FALSE
+	if(prob(50/severity))
```
Early-return

13spacemen committed:
early returns
improve malfunctioning proc
indent some text

SomeoneYouProbablyKnow commented:
I really like the "combat aspects" of it, running, dodging, etc, looks fun. But i think this is too good as a stealth tool. You keep brining up that it's unreliably but most of our walls are only 1 tile, and the % of a recharge will even out over time.
But i have a very simple solution, make the teleporter do a very obvious traitor-y sound (probably in both locations, the start and the end point). This way it keeps its uses as an obvious tool but not as a replacement to a sneaky tool like the briefcase. (Unless it already does that, the video doesnt have audio and im too lazy to check the code)

Additionally some sort of remnant (akin in function to emmagged airlocks) like a little portal fragment could be useful but at this point im just idea-guying.

Zevotech commented:
![image](https://user-images.githubusercontent.com/95449138/206079339-2b2d7376-2542-4327-a734-28df598c68d5.png)
please for the love of god nerf or rework this thing, its really funny but sec has no counterplay. my biggest threat was a beepsky chasing me for having no ID.

Mothblocks commented:
go on the forums
